### PR TITLE
ReplaceUserFunctions

### PR DESCRIPTION
**Pull request by Yahor-Zialeny**

### CODE REVIEW

File: Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs [hunk 1]
```diff
@@ -1,10 +1,49 @@
...
+    private InputPreprocessingService service = new InputPreprocessingService();
+    [Fact]
+    public void ReplaceUserFunctions_ValidInput_ReplacesFunctionCalls()
+    {
+        var functions = new Dictionary<string, UserFunction> {
+            { "f", new UserFunction { Name = "f", Variables = new List<string> { "x", "y" }, Expression = "x + y" } }
+        };
+        string input = "3 + f(1, 2)";
+        string result = service.ReplaceUserFunctions(input, functions);
+        Assert.Equal("3 + (1 + 2)", result);
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs:25]: Add blank lines to follow AAA pattern

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 2]
```diff
@@ -10,9 +10,31 @@ public class InputPreprocessingService
     /// </summary>
     /// <param name="input">Infix notation string with function calls</param>
     /// <returns>Infix notation string without function calls</returns>
-    public string ReplaceUserFunctions(string input)
-    {
-        throw new NotImplementedException();
+    public string ReplaceUserFunctions(string input, Dictionary<string, UserFunction> functions) {
+        foreach (var function in functions) {
+            string funcName = function.Key;
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: use var

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 3]
```diff
@@ -10,9 +10,31 @@ public class InputPreprocessingService
...
+            var userFunction = function.Value;
+            Regex regex = new Regex($@"\b{funcName}\(([^)]*)\)");
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: use var

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 4]
```diff
@@ -10,9 +10,31 @@ public class InputPreprocessingService
...
+            input = regex.Replace(input, match =>
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: Move to separated function

File: Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs [hunk 5]
```diff
@@ -1,10 +1,49 @@
...
+using System.Collections.Generic;
+using Calculator.Models;
+using Calculator.Services;
+
+
 public class InputPreprocessingServiceTests
 {
+    private InputPreprocessingService service = new InputPreprocessingService();
+    [Fact]
+    public void ReplaceUserFunctions_ValidInput_ReplacesFunctionCalls()
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs]: All test names must be in the same style and look like Subject_Action_ExpectedResult

**Review by Yahor-Zialeny — COMMENTED** [review]
All fixed

**Pull request merged**